subprojects/soak/src/integTest/groovy/org/gradle/vfs/VirtualFileSystemRetentionSoakTest.groovy

```diff
@@ -0,0 +1,150 @@
+        def rootProject = multiProjectBuild("javaProject", subprojects) {
+            buildFile << """
+                allprojects {
+                    apply plugin: 'java-library'
+
+                    tasks.withType(JavaCompile).configureEach {
+                        options.fork = true
+                    }
+                }
+            """
+        }
+        sourceFiles = subprojects.collectMany { projectDir ->
+            (1..NUMBER_OF_SOURCES_PER_SUBPROJECT).collect {
+                def sourceFile = rootProject.file("${projectDir}/src/main/java/my/domain/Dummy${it}.java")
+                modifySourceFile(sourceFile, 0)
+                return sourceFile
+            }
+        }
+
+        executer.beforeExecute {
+            withArgument("-D${VirtualFileSystemServices.VFS_RETENTION_ENABLED_PROPERTY}=true")
+        }
+    }
+
+    def "file watching works with multiple builds on the same daemon"() {
+        when:
+        succeeds("assemble")
+        def daemon = daemons.daemon
+        then:
+        daemon.assertIdle()
+
+        expect:
+        (1..100).each { iteration ->
+            changeAllSourceFiles(iteration)
+            waitForChangesToBePickedUp()
+            succeeds("assemble", "--parallel")
+            assert daemons.daemon.logFile == daemon.logFile
+            daemon.assertIdle()
+            assertWatchingSucceeded()
+            assert receivedFileSystemEvents >= expectedFileSystemEvents(TOTAL_NUMBER_OF_SOURCES, 1)
+        }
+    }
+
+    def "file watching works with many changes between two builds"() {
+        when:
+        succeeds("assemble", "--parallel")
+        def daemon = daemons.daemon
+        then:
+        daemon.assertIdle()
+
+        when:
+        (1..100).each { iteration ->
+            changeAllSourceFiles(iteration)
+            waitForChangesToBePickedUp()
+        }
+        then:
+        succeeds("assemble")
+        daemons.daemon.logFile == daemon.logFile
+        daemon.assertIdle()
+        assertWatchingSucceeded()
+        receivedFileSystemEvents >= expectedFileSystemEvents(TOTAL_NUMBER_OF_SOURCES, 100)
+    }
+
+    private static int expectedFileSystemEvents(int numberOfChangedFiles, int numberOfChangesPerFile) {
+        def currentOs = OperatingSystem.current()
+        if (currentOs.isMacOsX()) {
+            // macOS coalesces the changes if the are in short succession
+            return numberOfChangedFiles * numberOfChangesPerFile * LOST_EVENTS_RATIO_MAC_OS
+        } else if (currentOs.isLinux()) {
+            // the JDK watchers only capture one event per watched path
+            return numberOfChangedFiles
+        } else if (currentOs.isWindows()) {
+            return numberOfChangedFiles * numberOfChangesPerFile * LOST_EVENTS_RATIO_WINDOWS
+        }
+        throw new AssertionError("Test not supported on OS ${currentOs}")
+    }
+
+    private static void waitForChangesToBePickedUp() {
+        Thread.sleep(1000)
```

Comment:
We have a bunch of places where we now wait for a second for file system events to be picked up. It would be great to somehow centralize these, so later we can easily fine tune this time.

Maybe we should add a `VfsRetentionFixture`, similar to `DirectoryBuildCacheFixture` etc. We could also add something like `withVfsRetention()` in there to replace specifying the system property directly.

Reply:
Done.